onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperator.cpp

```diff
@@ -635,6 +635,40 @@ namespace Dml
             ));
     }
 
+    void DmlOperator::ConvertNHWCToNCHW(
+        const uint32_t& dimCount, 
+        const gsl::span<const uint32_t> nhwcSizes,
+        std::vector<uint32_t>& nchwSizes,
+        std::vector<uint32_t>& nchwInputStrides)
+    {
+        int i = 0;
+        const uint32_t inputBatch = nhwcSizes[ i++ ];
+        const uint32_t inputDepth = dimCount == 5 ? nhwcSizes[ i++ ] : 0;
```

Comment:
>dimCount == 5 [](http://example.com/codeflow?start=36&length=13)

`com.microsoft.QLinearAveragePool` can technically accept any size from 2D to ND, not just 4 or 5.

`For non image case, the dimensions are in the form of (N x C x D1 x D2 ... Dn), where N is the batch size.`

That means [N,C] and [N,C,D1,D2,D3,D4] are valid sizes.

Granted, DML itself doesn't accept arbitrary sizes, but the DML EP must be prepared for whatever ONNX can throw it and should be general enough so that when we update DML later, than the DML EP generally doesn't need updates too. See my comment in DmlOperatorQLinearAveragePool for a more generic function that obviates this one.

Reply:
Updated in commit 3